File: Akash-Job-Stuffs/Leetcode-Solve/Binary-Search/35. Search Insert Position.cpp
class Solution {
public:
    int searchInsert(vector<int>& nums, int target) {
        int n=nums.size();

        int low=0;
        int high=n-1;

        // we initialize "ans" with LENGTH of the array
        int ans=n;
        
        while(low<=high){
            int mid=low+((high-low)/2);

            if(nums[mid]==target){
                return mid;
            }

            // if "mid" is greater than target, that means we move the "high" pointer to the Right & "mid" COULD be our Answer, so we store it
            else if(target<nums[mid]){
                ans=mid;
                high=mid-1;
            }
            else{
                low=mid+1;
            }
        }

        return ans;
    }
};

File: Akash-Job-Stuffs/Leetcode-Solve/Binary-Search/35. Search Insert Position.java
class Solution {
    public int searchInsert(int[] nums, int target) {
        int low = 0;
        int high = nums.length-1;

        int ans = 0;

        while(low<=high){
            int mid  = low + ((high-low)/2);

            if(nums[mid]==target){
                return mid;
            }
            // After the End of the Loop, "low" will be the first index where nums[low] >= target, i.e., the place to insert target
            // this mean Target is on the Right
            else if(nums[mid]<target){
                low = mid+1;
                ans = low;
            }

            // this mean Target is on the Left
            else{
                high = mid-1;
            }
        }


        return ans;
    }
}